slither/utils/upgradeability.py

```diff
@@ -159,16 +174,137 @@ def compare(
+            ):
+                tainted_contracts[contract.name]["functions"].append(target)
+                for var in target.all_state_variables_written():
+                    if var not in tainted_contracts[contract.name]["variables"]:
+                        tainted_contracts[contract.name]["variables"].append(var)
+            elif (
+                isinstance(target, Variable)
+                and target not in tainted_contracts[contract.name]["variables"]
+                and not (target.is_constant or target.is_immutable)
+            ):
+                tainted_contracts[contract.name]["variables"].append(target)
+    for c in tainted_contracts.items():
+        if len(c[1]["functions"]) == 0 and len(c[1]["variables"]) == 0:
+            continue
+        tainted_list.append(c[1])
+        contract = c[1]["contract"]
+        variables = c[1]["variables"]
+        for var in variables:
+            read_write = set(
+                contract.get_functions_reading_from_variable(var)
+                + contract.get_functions_writing_to_variable(var)
+            )
+            for f in read_write:
+                if f not in tainted_contracts[contract.name]["functions"] and not (
+                    f.is_constructor or f.is_fallback or f.is_receive
+                ):
+                    tainted_contracts[contract.name]["functions"].append(f)
+    return tainted_list
+
+
+def tainted_inheriting_contracts(
+    tainted_contracts: List[TaintedExternalContract], contracts: List[Contract] = None
+) -> List[TaintedExternalContract]:
+    """
+    Takes a list of TaintedExternalContract obtained from tainted_external_contracts, and finds any contracts which
+    inherit a tainted contract, as well as any functions that call tainted functions or read tainted variables in
+    the inherited contract.
+    Args:
+        tainted_contracts: the list obtained from `tainted_external_contracts` or `compare`.
+        contracts: (optional) the list of contracts to check for inheritance. If not provided, defaults to
+                    `contract.compilation_unit.contracts` for each contract in tainted_contracts.
+
+    Returns:
+        An updated list of TaintedExternalContract, including all from the input list.
+    """
+    for tainted in tainted_contracts:
+        contract = tainted["contract"]
+        if contracts is None:
+            contracts = contract.compilation_unit.contracts
+        contracts = [
+            c
+            for c in contracts
+            if c.name not in tainted_contracts and c.name in [i.name for i in c.inheritance]
+        ]
+        for c in contracts:
+            new_taint = TaintedExternalContract(contract=c, functions=[], variables=[])
+            for f in c.functions_declared:
+                internal_calls = f.all_internal_calls()
+                if any(
+                    str(call) == str(t) for t in tainted["functions"] for call in internal_calls
```

Comment:
I think comparing a string here might lead to false result, as `str` will just return the name. Maybe we can compare the `canonical_name` instead?